diff --git a/openssl/crypto/evp/evp_enc.c b/openssl/crypto/evp/evp_enc.c
--- a/openssl/crypto/evp/evp_enc.c
+++ b/openssl/crypto/evp/evp_enc.c
@@ -1113,2 +1113,10 @@ int EVP_CIPHER_CTX_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg, void *ptr)
         break;
+    case EVP_CTRL_TLSTREE:
+        params[0] = OSSL_PARAM_construct_octet_string(OSSL_CIPHER_PARAM_TLSTREE, ptr, sz);
+        params[1] = OSSL_PARAM_construct_end();
+        break;
+    case EVP_CTRL_SET_TLSTREE_PARAMS:
+        params[0] = OSSL_PARAM_construct_octet_string(OSSL_CIPHER_PARAM_TLSTREE_MODE, ptr, sz);
+        params[1] = OSSL_PARAM_construct_end();
+        break;
     }
diff --git a/openssl/crypto/objects/obj_dat.h b/openssl/crypto/objects/obj_dat.h
--- a/openssl/crypto/objects/obj_dat.h
+++ b/openssl/crypto/objects/obj_dat.h
@@ -12,3 +12,3 @@
 /* Serialized OID's */
-static const unsigned char so[9582] = {
+static const unsigned char so[9600] = {
     0x2A,0x86,0x48,0x86,0xF7,0x0D,                 /* [    0] OBJ_rsadsi */
@@ -1356,5 +1356,7 @@ static const unsigned char so[9582] = {
     0x2A,0x86,0x48,0x86,0xF7,0x0D,0x01,0x09,0x10,0x03,0x11,  /* [ 9570] OBJ_id_alg_hss_lms_hashsig */
+    0x2A,0x85,0x03,0x07,0x01,0x01,0x05,0x01,0x03,  /* [ 9581] OBJ_magma_mgm */
+    0x2A,0x85,0x03,0x07,0x01,0x01,0x05,0x02,0x03,  /* [ 9590] OBJ_kuznyechik_mgm */
 };
 
-#define NUM_NID 1502
+#define NUM_NID 1504
 static const ASN1_OBJECT nid_objs[NUM_NID] = {
@@ -2862,5 +2864,7 @@ static const ASN1_OBJECT nid_objs[NUM_NID] = {
     {"id-alg-hss-lms-hashsig", "id-alg-hss-lms-hashsig", NID_id_alg_hss_lms_hashsig, 11, &so[9570]},
+    {"magma-mgm", "magma-mgm", NID_magma_mgm, 9, &so[9581]},
+    {"kuznyechik-mgm", "kuznyechik-mgm", NID_kuznyechik_mgm, 9, &so[9590]},
 };
 
-#define NUM_SN 1493
+#define NUM_SN 1495
 static const unsigned int sn_objs[NUM_SN] = {
@@ -3828,2 +3832,3 @@ static const unsigned int sn_objs[NUM_SN] = {
     1017,    /* "kuznyechik-mac" */
+    1503,    /* "kuznyechik-mgm" */
     1014,    /* "kuznyechik-ofb" */
@@ -3841,2 +3846,3 @@ static const unsigned int sn_objs[NUM_SN] = {
     1192,    /* "magma-mac" */
+    1502,    /* "magma-mgm" */
     1189,    /* "magma-ofb" */
@@ -4361,3 +4367,3 @@ static const unsigned int sn_objs[NUM_SN] = {
 
-#define NUM_LN 1493
+#define NUM_LN 1495
 static const unsigned int ln_objs[NUM_LN] = {
@@ -5441,2 +5447,3 @@ static const unsigned int ln_objs[NUM_LN] = {
     1017,    /* "kuznyechik-mac" */
+    1503,    /* "kuznyechik-mgm" */
     1014,    /* "kuznyechik-ofb" */
@@ -5466,2 +5473,3 @@ static const unsigned int ln_objs[NUM_LN] = {
     1192,    /* "magma-mac" */
+    1502,    /* "magma-mgm" */
     1189,    /* "magma-ofb" */
@@ -5858,3 +5866,3 @@ static const unsigned int ln_objs[NUM_LN] = {
 
-#define NUM_OBJ 1350
+#define NUM_OBJ 1352
 static const unsigned int obj_objs[NUM_OBJ] = {
@@ -6775,4 +6783,6 @@ static const unsigned int obj_objs[NUM_OBJ] = {
     1175,    /* OBJ_magma_ctr_acpkm_omac         1 2 643 7 1 1 5 1 2 */
+    1502,    /* OBJ_magma_mgm                    1 2 643 7 1 1 5 1 3 */
     1177,    /* OBJ_kuznyechik_ctr_acpkm         1 2 643 7 1 1 5 2 1 */
     1178,    /* OBJ_kuznyechik_ctr_acpkm_omac    1 2 643 7 1 1 5 2 2 */
+    1503,    /* OBJ_kuznyechik_mgm               1 2 643 7 1 1 5 2 3 */
     1181,    /* OBJ_magma_kexp15                 1 2 643 7 1 1 7 1 1 */
diff --git a/openssl/crypto/objects/obj_mac.num b/openssl/crypto/objects/obj_mac.num
--- a/openssl/crypto/objects/obj_mac.num
+++ b/openssl/crypto/objects/obj_mac.num
@@ -1501 +1501,3 @@ id_smime_ori_kem		1500
 id_alg_hss_lms_hashsig		1501
+magma_mgm		1502
+kuznyechik_mgm		1503
diff --git a/openssl/crypto/objects/objects.txt b/openssl/crypto/objects/objects.txt
--- a/openssl/crypto/objects/objects.txt
+++ b/openssl/crypto/objects/objects.txt
@@ -1531,2 +1531,3 @@ id-tc26-cipher-gostr3412-2015-magma 1	: magma-ctr-acpkm
 id-tc26-cipher-gostr3412-2015-magma 2	: magma-ctr-acpkm-omac
+id-tc26-cipher-gostr3412-2015-magma 3	: magma-mgm
 id-tc26-cipher 2	:  id-tc26-cipher-gostr3412-2015-kuznyechik
@@ -1534,2 +1535,3 @@ id-tc26-cipher-gostr3412-2015-kuznyechik 1	: kuznyechik-ctr-acpkm
 id-tc26-cipher-gostr3412-2015-kuznyechik 2	: kuznyechik-ctr-acpkm-omac
+id-tc26-cipher-gostr3412-2015-kuznyechik 3	: kuznyechik-mgm
 
diff --git a/openssl/fuzz/oids.txt b/openssl/fuzz/oids.txt
--- a/openssl/fuzz/oids.txt
+++ b/openssl/fuzz/oids.txt
@@ -1352 +1352,3 @@ OBJ_id_smime_ori_kem="\x2A\x86\x48\x86\xF7\x0D\x01\x09\x10\x0D\x03"
 OBJ_id_alg_hss_lms_hashsig="\x2A\x86\x48\x86\xF7\x0D\x01\x09\x10\x03\x11"
+OBJ_magma_mgm="\x2A\x85\x03\x07\x01\x01\x05\x01\x03"
+OBJ_kuznyechik_mgm="\x2A\x85\x03\x07\x01\x01\x05\x02\x03"
diff --git a/openssl/include/openssl/evp.h b/openssl/include/openssl/evp.h
--- a/openssl/include/openssl/evp.h
+++ b/openssl/include/openssl/evp.h
@@ -350,2 +350,3 @@ int EVP_default_properties_enable_fips(OSSL_LIB_CTX *libctx, int enable);
 #define EVP_CTRL_TLSTREE 0x2A
+#define EVP_CTRL_SET_TLSTREE_PARAMS 0x2B
 
@@ -388,2 +389,6 @@ typedef struct {
 
+/* GOST TLS 1.3 tag lengths */
+#define EVP_MAGMA_TLS_TAG_LEN 8
+#define EVP_KUZNYECHIK_TLS_TAG_LEN 16
+
 /* Length of tag for TLS */
diff --git a/openssl/include/openssl/obj_mac.h b/openssl/include/openssl/obj_mac.h
--- a/openssl/include/openssl/obj_mac.h
+++ b/openssl/include/openssl/obj_mac.h
@@ -4859,2 +4859,6 @@
 
+#define SN_magma_mgm            "magma-mgm"
+#define NID_magma_mgm           1502
+#define OBJ_magma_mgm           OBJ_id_tc26_cipher_gostr3412_2015_magma,3L
+
 #define SN_id_tc26_cipher_gostr3412_2015_kuznyechik             "id-tc26-cipher-gostr3412-2015-kuznyechik"
@@ -4871,2 +4875,6 @@
 
+#define SN_kuznyechik_mgm               "kuznyechik-mgm"
+#define NID_kuznyechik_mgm              1503
+#define OBJ_kuznyechik_mgm              OBJ_id_tc26_cipher_gostr3412_2015_kuznyechik,3L
+
 #define SN_id_tc26_agreement            "id-tc26-agreement"
diff --git a/openssl/include/openssl/tls1.h b/openssl/include/openssl/tls1.h
--- a/openssl/include/openssl/tls1.h
+++ b/openssl/include/openssl/tls1.h
@@ -676,2 +676,8 @@ int SSL_CTX_set_tlsext_ticket_key_evp_cb(SSL_CTX *ctx, int (*fp)(SSL *, unsigned
 
+/* TLS1.3 GOST ciphersuites from RFC9367 */
+#define TLS1_3_CK_GOSTR341112_256_WITH_KUZNYECHIK_MGM_L 0x0300C103
+#define TLS1_3_CK_GOSTR341112_256_WITH_MAGMA_MGM_L 0x0300C104
+#define TLS1_3_CK_GOSTR341112_256_WITH_KUZNYECHIK_MGM_S 0x0300C105
+#define TLS1_3_CK_GOSTR341112_256_WITH_MAGMA_MGM_S 0x0300C106
+
 /* SM ciphersuites from RFC8998 */
@@ -872,2 +878,6 @@ int SSL_CTX_set_tlsext_ticket_key_evp_cb(SSL_CTX *ctx, int (*fp)(SSL *, unsigned
 #define TLS1_RFC_RSA_PSK_WITH_ARIA_256_GCM_SHA384 "TLS_RSA_PSK_WITH_ARIA_256_GCM_SHA384"
+#define TLS1_RFC_GOSTR341112_256_WITH_KUZNYECHIK_MGM_L "TLS_GOSTR341112_256_WITH_KUZNYECHIK_MGM_L"
+#define TLS1_RFC_GOSTR341112_256_WITH_KUZNYECHIK_MGM_S "TLS_GOSTR341112_256_WITH_KUZNYECHIK_MGM_S"
+#define TLS1_RFC_GOSTR341112_256_WITH_MAGMA_MGM_L "TLS_GOSTR341112_256_WITH_MAGMA_MGM_L"
+#define TLS1_RFC_GOSTR341112_256_WITH_MAGMA_MGM_S "TLS_GOSTR341112_256_WITH_MAGMA_MGM_S"
 #define TLS1_3_RFC_SM4_GCM_SM3 "TLS_SM4_GCM_SM3"
diff --git a/openssl/ssl/record/methods/recmethod_local.h b/openssl/ssl/record/methods/recmethod_local.h
--- a/openssl/ssl/record/methods/recmethod_local.h
+++ b/openssl/ssl/record/methods/recmethod_local.h
@@ -333,3 +333,3 @@ struct ossl_record_layer_st {
     int stream_mac;
-    int tlstree;
+    uint32_t tlstree;
 
diff --git a/openssl/ssl/record/methods/tls13_meth.c b/openssl/ssl/record/methods/tls13_meth.c
--- a/openssl/ssl/record/methods/tls13_meth.c
+++ b/openssl/ssl/record/methods/tls13_meth.c
@@ -82,2 +82,14 @@ static int tls13_set_crypto_state(OSSL_RECORD_LAYER *rl, int level,
     }
+
+    if (!rl->isdtls && rl->tlstree) {
+        int res = 0;
+
+        if ((rl->tlstree & TLS1_TLSTREE_S) != 0)
+            res = EVP_CIPHER_CTX_ctrl(ciph_ctx, EVP_CTRL_SET_TLSTREE_PARAMS, 0, "strong");
+        else if ((rl->tlstree & TLS1_TLSTREE_L) != 0)
+            res = EVP_CIPHER_CTX_ctrl(ciph_ctx, EVP_CTRL_SET_TLSTREE_PARAMS, 0, "light");
+
+        if (res <= 0)
+            return OSSL_RECORD_RETURN_FATAL;
+    }
 end:
@@ -166,2 +178,7 @@ static int tls13_cipher(OSSL_RECORD_LAYER *rl, TLS_RL_RECORD *recs,
 
+    if (!rl->isdtls && rl->tlstree && EVP_CIPHER_CTX_ctrl(enc_ctx, EVP_CTRL_TLSTREE, 0, seq) <= 0) {
+        RLAYERfatal(rl, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+
     if (!tls_increment_sequence_ctr(rl)) {
diff --git a/openssl/ssl/record/methods/tls_common.c b/openssl/ssl/record/methods/tls_common.c
--- a/openssl/ssl/record/methods/tls_common.c
+++ b/openssl/ssl/record/methods/tls_common.c
@@ -1258,3 +1258,3 @@ int tls_int_new_record_layer(OSSL_LIB_CTX *libctx, const char *propq, int vers,
                 == 0) {
-                if (!OSSL_PARAM_get_int(p, &rl->tlstree)) {
+                if (!OSSL_PARAM_get_uint32(p, &rl->tlstree)) {
                     ERR_raise(ERR_LIB_SSL, SSL_R_FAILED_TO_GET_PARAMETER);
diff --git a/openssl/ssl/record/rec_layer_s3.c b/openssl/ssl/record/rec_layer_s3.c
--- a/openssl/ssl/record/rec_layer_s3.c
+++ b/openssl/ssl/record/rec_layer_s3.c
@@ -1253,3 +1253,4 @@ int ssl_set_new_record_layer(SSL_CONNECTION *s, int version,
     const OSSL_RECORD_METHOD *meth;
-    int use_etm, stream_mac = 0, tlstree = 0;
+    int use_etm, stream_mac = 0;
+    uint32_t algorithm2 = 0, tlstree = 0;
     unsigned int maxfrag = (direction == OSSL_RECORD_DIRECTION_WRITE)
@@ -1316,2 +1317,10 @@ int ssl_set_new_record_layer(SSL_CONNECTION *s, int version,
 
+    if (s->s3.tmp.new_cipher != NULL)
+        algorithm2 = s->s3.tmp.new_cipher->algorithm2;
+    else if (s->session != NULL && s->session->cipher != NULL)
+        algorithm2 = s->session->cipher->algorithm2;
+    else if (s->psksession != NULL && s->psksession->cipher != NULL)
+        algorithm2 = s->psksession->cipher->algorithm2;
+    tlstree |= (algorithm2 & TLS1_TLSTREE_L) | (algorithm2 & TLS1_TLSTREE_S);
+
     if (use_etm)
@@ -1325,4 +1334,3 @@ int ssl_set_new_record_layer(SSL_CONNECTION *s, int version,
     if (tlstree)
-        *set++ = OSSL_PARAM_construct_int(OSSL_LIBSSL_RECORD_LAYER_PARAM_TLSTREE,
-            &tlstree);
+        *set++ = OSSL_PARAM_construct_uint32(OSSL_LIBSSL_RECORD_LAYER_PARAM_TLSTREE, &tlstree);
 
diff --git a/openssl/ssl/s3_lib.c b/openssl/ssl/s3_lib.c
--- a/openssl/ssl/s3_lib.c
+++ b/openssl/ssl/s3_lib.c
@@ -204,2 +204,76 @@ static SSL_CIPHER tls13_ciphers[] = {
     },
+#ifndef OPENSSL_NO_GOST
+    {
+        1,
+        TLS1_RFC_GOSTR341112_256_WITH_KUZNYECHIK_MGM_L,
+        TLS1_RFC_GOSTR341112_256_WITH_KUZNYECHIK_MGM_L,
+        TLS1_3_CK_GOSTR341112_256_WITH_KUZNYECHIK_MGM_L,
+        SSL_kANY,
+        SSL_aANY,
+        SSL_KUZNYECHIK_MGM,
+        SSL_AEAD,
+        TLS1_3_VERSION,
+        TLS1_3_VERSION,
+        0,
+        0,
+        SSL_HIGH,
+        SSL_HANDSHAKE_MAC_GOST12_256 | TLS1_PRF_GOST12_256 | TLS1_TLSTREE | TLS1_TLSTREE_L,
+        256,
+        256,
+    },
+    {
+        1,
+        TLS1_RFC_GOSTR341112_256_WITH_KUZNYECHIK_MGM_S,
+        TLS1_RFC_GOSTR341112_256_WITH_KUZNYECHIK_MGM_S,
+        TLS1_3_CK_GOSTR341112_256_WITH_KUZNYECHIK_MGM_S,
+        SSL_kANY,
+        SSL_aANY,
+        SSL_KUZNYECHIK_MGM,
+        SSL_AEAD,
+        TLS1_3_VERSION,
+        TLS1_3_VERSION,
+        0,
+        0,
+        SSL_HIGH,
+        SSL_HANDSHAKE_MAC_GOST12_256 | TLS1_PRF_GOST12_256 | TLS1_TLSTREE | TLS1_TLSTREE_S,
+        256,
+        256,
+    },
+    {
+        1,
+        TLS1_RFC_GOSTR341112_256_WITH_MAGMA_MGM_L,
+        TLS1_RFC_GOSTR341112_256_WITH_MAGMA_MGM_L,
+        TLS1_3_CK_GOSTR341112_256_WITH_MAGMA_MGM_L,
+        SSL_kANY,
+        SSL_aANY,
+        SSL_MAGMA_MGM,
+        SSL_AEAD,
+        TLS1_3_VERSION,
+        TLS1_3_VERSION,
+        0,
+        0,
+        SSL_HIGH,
+        SSL_HANDSHAKE_MAC_GOST12_256 | TLS1_PRF_GOST12_256 | TLS1_TLSTREE | TLS1_TLSTREE_L,
+        256,
+        256,
+    },
+    {
+        1,
+        TLS1_RFC_GOSTR341112_256_WITH_MAGMA_MGM_S,
+        TLS1_RFC_GOSTR341112_256_WITH_MAGMA_MGM_S,
+        TLS1_3_CK_GOSTR341112_256_WITH_MAGMA_MGM_S,
+        SSL_kANY,
+        SSL_aANY,
+        SSL_MAGMA_MGM,
+        SSL_AEAD,
+        TLS1_3_VERSION,
+        TLS1_3_VERSION,
+        0,
+        0,
+        SSL_HIGH,
+        SSL_HANDSHAKE_MAC_GOST12_256 | TLS1_PRF_GOST12_256 | TLS1_TLSTREE | TLS1_TLSTREE_S,
+        256,
+        256,
+    },
+#endif
 };
diff --git a/openssl/ssl/ssl_ciph.c b/openssl/ssl/ssl_ciph.c
--- a/openssl/ssl/ssl_ciph.c
+++ b/openssl/ssl/ssl_ciph.c
@@ -65,2 +65,4 @@ static const ssl_cipher_table ssl_cipher_table_cipher[SSL_ENC_NUM_IDX] = {
     { SSL_SM4CCM, NID_sm4_ccm }, /* SSL_ENC_SM4CCM_IDX 25 */
+    { SSL_MAGMA_MGM, NID_magma_mgm }, /* SSL_ENC_MAGMA_MGM_IDX 26 */
+    { SSL_KUZNYECHIK_MGM, NID_kuznyechik_mgm }, /* SSL_ENC_KUZNYECHIK_MGM_IDX 27 */
 };
@@ -276,2 +278,18 @@ static const SSL_CIPHER cipher_aliases[] = {
 
+#define IS_FETCHABLE(type, TYPE)                                         \
+    static int is_ ## type ## _fetchable(SSL_CTX *ctx, const char *name) \
+    {                                                                    \
+        TYPE *impl;                                                      \
+                                                                         \
+        ERR_set_mark();                                                  \
+        impl = TYPE ## _fetch(ctx->libctx, name, ctx->propq);            \
+        ERR_pop_to_mark();                                               \
+        if (impl == NULL)                                                \
+            return 0;                                                    \
+        TYPE ## _free(impl);                                             \
+        return 1;                                                        \
+    }
+
+IS_FETCHABLE(keymgmt, EVP_KEYMGMT)
+
 int ssl_load_ciphers(SSL_CTX *ctx)
@@ -388,5 +406,8 @@ int ssl_load_ciphers(SSL_CTX *ctx)
 
-    ctx->disabled_auth_mask |= SSL_aGOST01 | SSL_aGOST12;
-    ctx->disabled_auth_mask |= SSL_aGOST12;
-    ctx->disabled_auth_mask |= SSL_aGOST12;
+    if (!is_keymgmt_fetchable(ctx, SN_id_GostR3410_2001))
+        ctx->disabled_auth_mask |= SSL_aGOST01 | SSL_aGOST12;
+    if (!is_keymgmt_fetchable(ctx, SN_id_GostR3410_2012_256))
+        ctx->disabled_auth_mask |= SSL_aGOST12;
+    if (!is_keymgmt_fetchable(ctx, SN_id_GostR3410_2012_512))
+        ctx->disabled_auth_mask |= SSL_aGOST12;
     /*
@@ -1790,2 +1811,5 @@ char *SSL_CIPHER_description(const SSL_CIPHER *cipher, char *buf, int len)
         break;
+    case SSL_MAGMA_MGM:
+        enc = "MAGMAMGM";
+        break;
     case SSL_KUZNYECHIK:
@@ -1793,2 +1817,5 @@ char *SSL_CIPHER_description(const SSL_CIPHER *cipher, char *buf, int len)
         break;
+    case SSL_KUZNYECHIK_MGM:
+        enc = "KUZNYECHIKMGM";
+        break;
     case SSL_CHACHA20POLY1305:
diff --git a/openssl/ssl/ssl_local.h b/openssl/ssl/ssl_local.h
--- a/openssl/ssl/ssl_local.h
+++ b/openssl/ssl/ssl_local.h
@@ -159,2 +159,4 @@
 #define SSL_SM4CCM 0x02000000U
+#define SSL_MAGMA_MGM 0x04000000U
+#define SSL_KUZNYECHIK_MGM 0x08000000U
 
@@ -244,2 +246,4 @@
 #define TLS1_TLSTREE 0x20000
+#define TLS1_TLSTREE_S 0x80000
+#define TLS1_TLSTREE_L 0x100000
 
@@ -363,3 +367,5 @@
 #define SSL_ENC_SM4CCM_IDX 25
-#define SSL_ENC_NUM_IDX 26
+#define SSL_ENC_MAGMA_MGM_IDX 26
+#define SSL_ENC_KUZNYECHIK_MGM_IDX 27
+#define SSL_ENC_NUM_IDX 28
 
diff --git a/openssl/ssl/statem/statem_lib.c b/openssl/ssl/statem/statem_lib.c
--- a/openssl/ssl/statem/statem_lib.c
+++ b/openssl/ssl/statem/statem_lib.c
@@ -499,6 +499,6 @@ MSG_PROCESS_RETURN tls_process_cert_verify(SSL_CONNECTION *s, PACKET *pkt)
         && ((PACKET_remaining(pkt) == 64
-                && (EVP_PKEY_get_id(pkey) == NID_id_GostR3410_2001
-                    || EVP_PKEY_get_id(pkey) == NID_id_GostR3410_2012_256))
+                && (EVP_PKEY_is_a(pkey, SN_id_GostR3410_2001)
+                    || EVP_PKEY_is_a(pkey, SN_id_GostR3410_2012_256)))
             || (PACKET_remaining(pkt) == 128
-                && EVP_PKEY_get_id(pkey) == NID_id_GostR3410_2012_512))) {
+                && EVP_PKEY_is_a(pkey, SN_id_GostR3410_2012_512)))) {
         len = (unsigned int)PACKET_remaining(pkt);
@@ -547,6 +547,5 @@ MSG_PROCESS_RETURN tls_process_cert_verify(SSL_CONNECTION *s, PACKET *pkt)
     {
-        int pktype = EVP_PKEY_get_id(pkey);
-        if (pktype == NID_id_GostR3410_2001
-            || pktype == NID_id_GostR3410_2012_256
-            || pktype == NID_id_GostR3410_2012_512) {
+        if (EVP_PKEY_is_a(pkey, SN_id_GostR3410_2001)
+            || EVP_PKEY_is_a(pkey, SN_id_GostR3410_2012_256)
+            || EVP_PKEY_is_a(pkey, SN_id_GostR3410_2012_512)) {
             if ((gost_data = OPENSSL_malloc(len)) == NULL)
@@ -1904,4 +1903,2 @@ static int is_tls13_capable(const SSL_CONNECTION *s)
         case SSL_PKEY_GOST01:
-        case SSL_PKEY_GOST12_256:
-        case SSL_PKEY_GOST12_512:
             continue;
@@ -1922,2 +1919,11 @@ static int is_tls13_capable(const SSL_CONNECTION *s)
             return 1;
+
+        /*
+         * TODO: There must be an opportunity to check that the sig algs are
+         * consistent with the keys in case if the sig algs are provider-based.
+         * For example, OSSL_FUNC_keymgmt_get_params could return the list of
+         * sig alg code points supported by the key, and this list could be
+         * checked against the sig algs code points stored in
+         * SSL_CTX.sigalg_lookup_cache
+         */
     }
diff --git a/openssl/ssl/t1_lib.c b/openssl/ssl/t1_lib.c
--- a/openssl/ssl/t1_lib.c
+++ b/openssl/ssl/t1_lib.c
@@ -2372,3 +2372,8 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
     for (i = 0; i < ctx->sigalg_list_len; i++) {
+        size_t idx;
         TLS_SIGALG_INFO si = ctx->sigalg_list[i];
+
+        if (!ssl_cert_lookup_by_nid(OBJ_txt2nid(si.sigalg_name), &idx, ctx))
+            goto err;
+
         cache[cache_idx].name = si.name;
@@ -2380,3 +2385,3 @@ int ssl_setup_sigalgs(SSL_CTX *ctx)
         cache[cache_idx].sig = OBJ_txt2nid(si.sigalg_name);
-        cache[cache_idx].sig_idx = (int)(i + SSL_PKEY_NUM);
+        cache[cache_idx].sig_idx = (int)idx;
         cache[cache_idx].sigandhash = OBJ_txt2nid(si.sigalg_name);
@@ -3497,6 +3502,4 @@ static int tls12_sigalg_allowed(const SSL_CONNECTION *s, int op,
 
-    if (lu->sig == NID_id_GostR3410_2012_256
-        || lu->sig == NID_id_GostR3410_2012_512
-        || lu->sig == NID_id_GostR3410_2001) {
-        /* We never allow GOST sig algs on the server with TLSv1.3 */
+    if (lu->sig == NID_id_GostR3410_2001) {
+        /* We never allow GOST2001 sig algs on the server with TLSv1.3 */
         if (s->server && SSL_CONNECTION_IS_TLS13(s))
@@ -3536,2 +3539,14 @@ static int tls12_sigalg_allowed(const SSL_CONNECTION *s, int op,
 
+    if ((lu->sig == NID_id_GostR3410_2012_256
+         || lu->sig == NID_id_GostR3410_2012_512)
+        && ((s->server && SSL_CONNECTION_IS_TLS13(s))
+            || (!s->server
+                && SSL_CONNECTION_GET_SSL(s)->method->version == TLS_ANY_VERSION
+                && s->s3.tmp.min_ver >= TLS1_3_VERSION))
+        && (strcmp(lu->name, TLSEXT_SIGALG_gostr34102012_256_intrinsic_name) == 0
+            || strcmp(lu->name, TLSEXT_SIGALG_gostr34102012_512_intrinsic_name) == 0
+            || strcmp(lu->name, TLSEXT_SIGALG_gostr34102012_256_gostr34112012_256_name) == 0
+            || strcmp(lu->name, TLSEXT_SIGALG_gostr34102012_512_gostr34112012_512_name) == 0))
+        return 0;
+
     /* Finally see if security callback allows it */
diff --git a/openssl/ssl/tls13_enc.c b/openssl/ssl/tls13_enc.c
--- a/openssl/ssl/tls13_enc.c
+++ b/openssl/ssl/tls13_enc.c
@@ -393,2 +393,3 @@ static int derive_secret_key_and_iv(SSL_CONNECTION *s, const EVP_MD *md,
     } else {
+        uint32_t algenc;
 
@@ -396,19 +397,18 @@ static int derive_secret_key_and_iv(SSL_CONNECTION *s, const EVP_MD *md,
 
+        if (s->s3.tmp.new_cipher != NULL) {
+            algenc = s->s3.tmp.new_cipher->algorithm_enc;
+        } else if (s->session->cipher != NULL) {
+            /* We've not selected a cipher yet - we must be doing early data */
+            algenc = s->session->cipher->algorithm_enc;
+        } else if (s->psksession != NULL && s->psksession->cipher != NULL) {
+            /* We must be doing early data with out-of-band PSK */
+            algenc = s->psksession->cipher->algorithm_enc;
+        } else {
+            SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_EVP_LIB);
+            return 0;
+        }
+
         mode = EVP_CIPHER_get_mode(ciph);
         if (mode == EVP_CIPH_CCM_MODE) {
-            uint32_t algenc;
-
             *ivlen = EVP_CCM_TLS_IV_LEN;
-            if (s->s3.tmp.new_cipher != NULL) {
-                algenc = s->s3.tmp.new_cipher->algorithm_enc;
-            } else if (s->session->cipher != NULL) {
-                /* We've not selected a cipher yet - we must be doing early data */
-                algenc = s->session->cipher->algorithm_enc;
-            } else if (s->psksession != NULL && s->psksession->cipher != NULL) {
-                /* We must be doing early data with out-of-band PSK */
-                algenc = s->psksession->cipher->algorithm_enc;
-            } else {
-                SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_EVP_LIB);
-                return 0;
-            }
             if (algenc & (SSL_AES128CCM8 | SSL_AES256CCM8))
@@ -422,2 +422,8 @@ static int derive_secret_key_and_iv(SSL_CONNECTION *s, const EVP_MD *md,
                 *taglen = EVP_GCM_TLS_TAG_LEN;
+#ifndef OPENSSL_NO_GOST
+            } else if ((algenc & SSL_MAGMA_MGM) != 0) {
+                *taglen = EVP_MAGMA_TLS_TAG_LEN;
+            } else if ((algenc & SSL_KUZNYECHIK_MGM) != 0) {
+                *taglen = EVP_KUZNYECHIK_TLS_TAG_LEN;
+#endif
             } else {
diff --git a/openssl/util/perl/OpenSSL/paramnames.pm b/openssl/util/perl/OpenSSL/paramnames.pm
--- a/openssl/util/perl/OpenSSL/paramnames.pm
+++ b/openssl/util/perl/OpenSSL/paramnames.pm
@@ -156,2 +156,4 @@ my %params = (
     'OSSL_CIPHER_HMAC_PARAM_MAC' =>             "*OSSL_CIPHER_PARAM_AEAD_TAG",
+    'OSSL_CIPHER_PARAM_TLSTREE' =>              "tlstree",             # octet_string
+    'OSSL_CIPHER_PARAM_TLSTREE_MODE' =>         "tlstree_mode",        # octet_string
 
